### PR TITLE
[ModuloSchedule] Implement modulo variable expansion for pipelining

### PR DESCRIPTION
**Pull request by ytmukai**
Modulo variable expansion is a technique that resolves overlap of variable lifetimes by unrolling. The existing implementation solves it by making a copy by move instruction for processors with ordinary registers such as Arm and x86. This method may result in a very large number of move instructions, which can cause performance problems.

Modulo variable expansion is enabled by specifying -pipeliner-mve-cg. A backend must implement some newly defined interfaces in PipelinerLoopInfo.

Discourse thread: https://discourse.llvm.org/t/implementing-modulo-variable-expansion-for-machinepipeliner

### CODE REVIEW

**Comment by ytmukai** [comment]
I also created an implementation for AArch64 at https://github.com/ytmukai/llvm-project/tree/swpl-aarch64. I will create a PR for it after this PR is merged, as it depends on this PR.

I measured the effect of modulo variable expansion (MVE) with llvm-test-suite. I used Ampere Altra Max, an AArch64 processor with Neoverse N1 core.
The performance compared to the non-pipelined results are as follows:

| MVE | >3% improvement cases | >3% degradation cases |
| --- | --- | --- |
| Disabled (conventional) | 50 | 222 |
| Enabled | 71 | 115 |

The total number of test cases is 811, selected by enabling TEST_SUITE_BENCHMARKING_ONLY.

The number of cases in which the performance improvement is more than 3% has increased by 40%.
There are still many cases showing degradation. I think this could be improved by adding a process to limit register pressure.

The optimization flags used are as follows: 
* non-pipelined: -O3
* MVE disabled: non-pipelined flags +  -mllvm -aarch64-enable-pipeliner=1 -mllvm -pipeliner-max-stages=1000 -mllvm --pipeliner-max-mii=1000 -mllvm -enable-misched=0 -mllvm -enable-post-misched=0 -mtune=neoverse-n1 -mllvm -pipeliner-enable-copytophi=0
* MVE enabled: MVE disabled flags + -mllvm -pipeliner-mve-cg=1

**Comment by ytmukai** [comment]
ping

File: llvm/lib/CodeGen/ModuloSchedule.cpp
Comment (DragonDisciple) [inline llvm/lib/CodeGen/ModuloSchedule.cpp:2673]: Would be a good debug enhancement to add small explanations to each failure so users know what in particular causes this not to apply.

File: llvm/lib/CodeGen/ModuloSchedule.cpp
Comment (DragonDisciple) [inline llvm/lib/CodeGen/ModuloSchedule.cpp:2689]: Side note: I'm curious how frequently and why LLVM generates PHI -> PHI within the same loop. I had only first seen it happen in the default ModuloScheduleExpander, which uses same-loop PHI->PHI lifetimes to represent values being passed through the pipeline stages.

What I'm asking I suppose is, was there an error case where you saw this occur?

File: llvm/lib/CodeGen/ModuloSchedule.cpp
Comment (ytmukai) [reply llvm/lib/CodeGen/ModuloSchedule.cpp]: Thanks, updated as 85175fa692f8f7668f9bf46e8245f7fdff6da39c

File: llvm/lib/CodeGen/ModuloSchedule.cpp
Comment (ytmukai) [reply llvm/lib/CodeGen/ModuloSchedule.cpp:2689]: It seems that such phi may be generated as a result of eliminating mov as follows.
https://llvm.godbolt.org/z/3sMqc5hTK

File: llvm/include/llvm/CodeGen/TargetInstrInfo.h
Comment (ytmukai) [reply llvm/include/llvm/CodeGen/TargetInstrInfo.h]: Thanks for the advice! I fixed all points with 354a42c18c87f302a2b81c07e9bc026ea342e2d3: rename and integration to createRemainingIterationsGreaterCondition(), addition of isMVESupported(), non-use of default implementation. I verified that the tests for the existing target (Hexagon, ARM and PowerPC) pass.

**Comment by hgreving2304** [comment]
> ping

Cool.

Could you describe this a bit more (for example how are low trip counts handled?)

I'm actually not sure if the upstream LLVM swing scheduler stitches the prolig/epilog schedule for low trip counts or whether it versions the loop.

Could you comment on the cost model used to unroll per MVE? It should only be needed if there are copies, or an excessive number of copies.

File: llvm/lib/CodeGen/ModuloSchedule.cpp
Comment (hgreving2304) [reply llvm/lib/CodeGen/ModuloSchedule.cpp:2689]: You'd always have a phi-phi if there is a loop-carried value with input distance > 1, in which case during software pipelining, the producer can move into a higher stage than the consumer (not sure if LLVM's upstream modulo scheduler can support this, we can in our downstream version). During swp, you'd always have phi-phi if producer stage - consumer stage > 1

**Comment by ytmukai** [comment]
In order to verify the effect of MVE in more detail, I tested it in a scientific computation loop, for which software pipelining would be useful. The tested code is the following loop from [IAMR](https://proxyapps.exascaleproject.org/app/iamr/), one of the ECP proxy applications, with some modifications.
https://github.com/AMReX-Codes/amrex/blob/9e35dc19489dc5d312e92781cb0471d282cf8370/Src/LinearSolvers/MLMG/AMReX_MLNodeLap_2D_K.H#L584
(The modifications include removing branches and adjusting the number of iterations so that accesses hit the L1 cache to make it easier to check the effects of the optimization, which may change the conditions from the actual application execution. I am working on making the modified code publicly available.)

The modified code was measured with Graviton3 (Neoverse V1) and the results are as follows.

| -pipeliner-force-ii=N | cycles (no mve) | instructions (no mve) | cycles (mve) | instructions (mve) | cycles (no swpl) | instructions (no swpl) |
| --------------------: | --------------: | --------------------: | -----------: | -----------------: | ---------------: | ---------------------: |
|                  - |               -  |                    -   |         -     |               -     |             18.0 |                   62.6 |
|              11 (MII) |            29.3 |                 134.4 |         19.6 |              100.7 |                - |                      - |
|                    12 |            22.1 |                 119.0 |         18.1 |               95.1 |                - |                      - |
|                    13 |            20.0 |                 103.8 |         17.6 |               89.8 |                - |                      - |
|                    14 |            17.8 |                  92.1 |         17.1 |               86.4 |                - |                      - |
|                    15 |            17.3 |                  85.3 |         16.6 |               80.4 |                - |                      - |
|                    16 |            18.4 |                  89.1 |         16.4 |               75.0 |                - |                      - |
|                    17 |            17.5 |                  80.3 |         15.6 |               67.6 |                - |                      - |
|                    18 |            17.2 |                  76.4 |         15.7 |               65.6 |                - |                      - |
|                    19 |            16.6 |                  70.4 |     __15.5__ |               64.1 |                - |                      - |
|                    20 |            16.5 |                  69.5 |         15.7 |               64.6 |                - |                      - |
|                    21 |            16.6 |                  69.5 |         15.7 |               64.6 |                - |                      - |
|                    22 |        __15.9__ |                  67.6 |         15.7 |               64.6 |                - |                      - |
|                    23 |            16.5 |                  67.5 |         16.1 |               64.6 |                - |                      - |
|                    24 |            16.5 |                  67.6 |         16.0 |               64.6 |                - |                      - |
|                    25 |            16.5 |                  66.5 |         16.2 |               64.6 |                - |                      - |

The table compares the number of cycles and instructions per iteration with and without MVE for each scheduled II. The number of cycles is reduced by 3% by MVE when comparing the fastest results for each. The reduction is 14% compared to the case without pipelining.

The number of floating-point instructions per iteration of the loop is 43 (division is reciprocally approximated by -mrecip) and Neoverse V1 has four arithmetic units, so ResMII is 11. With small II, the performance is reduced by spills due to lack of register.
 This result indicates that the number of registers should also be considered when determining II. My colleague @kasuga-fj is working on this issue (https://discourse.llvm.org/t/considering-register-pressure-when-deciding-initiation-interval-in-machinepipeliner/74725).

I would appreciate any advice for further verification.

**Comment by ytmukai** [comment]
I have updated the patch; interface fixes and AArch64 implementation. Could you please restart the review?

* Replace a parameter `RegMap` to `LastStage0Insts` in `createRemainingIterationsGreaterCondition()`. These maps are used to obtain the loop counter value at that point. We found that `RegMap` is not sufficient if that register is defined by PHI, so it was changed to an instruction map.
* Implement interfaces of MVE for AArch64. The main part is to determine whether the loop ends at the unrolled kernel. It is implemented in a way that generates comparison instructions for the next unroll count iteration and accumulates the results with CINC instructions. Although the number of instructions is large, it can handle many forms of loops and is not a significant problem for large kernels where pipelining is effective.

I tested the performance with llvm-test-suite on a Neoverse V1 processor. The following table lists the most effective test cases with pipelining. All results are uploaded as [swp.performance.llvm-test-suite.csv](https://github.com/llvm/llvm-project/files/14960352/swp.performance.llvm-test-suite.csv).

| test case                                                        | NOSWP |  SWP | SWP+MVE | SWP+faster | speedup by SWP(faster) | speedup by MVE |
| ---------------------------------------------------------------- | ----: | ---: | ------: | ---------: | ---------------------: | -------------: |
| SingleSource/Benchmarks/Misc/flops-3.test                        |  0.67 | 0.82 |    0.54 |       0.54 |                  24.9% |          53.2% |
| MultiSource/Benchmarks/TSVC/Recurrences-flt/Recurrences-flt.test |  2.59 | 2.61 |    2.24 |       2.24 |                  15.6% |          16.5% |
| MultiSource/Benchmarks/TSVC/Recurrences-dbl/Recurrences-dbl.test |  2.62 | 2.63 |    2.28 |       2.28 |                  14.7% |          15.1% |
| SingleSource/Benchmarks/Misc/flops-1.test                        |  0.54 | 0.68 |    0.47 |       0.47 |                  13.6% |          43.1% |
| SingleSource/Benchmarks/Adobe-C++/loop_unroll.test               |  0.46 | 0.41 |    0.41 |       0.41 |                  12.7% |           0.0% |
| SingleSource/Benchmarks/Misc/flops.test                          |  2.66 | 2.79 |    2.41 |       2.41 |                  10.3% |          15.7% |

Optimization flags:
* NOSWP: -O3 -mcpu=neoverse-v1
* SWP: NOSWP + -mllvm -aarch64-enable-pipeliner -mllvm -pipeliner-max-stages=100 -mllvm -pipeliner-max-mii=100 -mllvm -pipeliner-enable-copytophi=0 -mllvm -pipeliner-register-pressure
* SWP+MVE: SWP + -mllvm -pipeliner-mve-cg
* SWP+faster is the faster one for each cases.

In many of the above cases, the effect of MVE is significant and the csv shows that it is better in general. I believe that MVE should be the default in AArch64.

Note: Cases with unstable runtimes and micro-benchmarks are excluded. The micro-benchmarks are mostly the same and are not suitable for pipelining because its kernel is very small.

**Comment by ytmukai** [comment]
We have also evaluated other benchmarks and will share the results.

In a hotspot kernel in [ExaMiniMD](https://github.com/ECP-copa/ExaMiniMD), pipelining with MVE is effective.

ExaMiniMD is a benchmark for molecular dynamics in [ECP Proxy Applications](https://proxyapps.exascaleproject.org/). Then the [kernel](https://github.com/ECP-copa/ExaMiniMD/blob/3264e29e28e7a5a4695a959c12df6bdf03bada34/src/force_types/force_lj_neigh_impl.h#L178) calculates the LJ potential, a kind of pair potential.

It was measured on one Neoverse V1 core. The execution time of the kernel accounts for approximately 50% of the total. The number of cycles of the kernel measured by perf record is as follows (The fastest II is selected for each):

| NOSWP | SWP(II=25) | SWP+MVE(II=17) |
| --- | --- | --- |
| 15.2e9 | 15.3e9 | 14.2e9 |

(Compile flags: -Ofast -mrecip -g -mcpu=neoverse-v1 -mllvm -sve-gather-overhead=1 -mllvm -sve-scatter-overhead=1 -mllvm -aarch64-enable-pipeliner=1 -mllvm -pipeliner-max-stages=100 -mllvm -pipeliner-max-mii=100 -mllvm -pipeliner-enable-copytophi=0 -mllvm -pipeliner-mve-cg=0or1 -mllvm -pipeliner-force-ii=N -fno-unroll-loops)

Pipelining with MVE reduced execution time by 6.7%. (3.4% in total)

We also evaluated other benchmarks in ECP Proxy Applications, CORAL-2 Benchmarks and SPEC, but no hotspot kernels found to be effective. This was mostly due to the fact that pipelining was inherently ineffective for the kernels because of a small number instructions, etc. Some kernels could benefit if pipelining can be applied by improving other optimizations, alias analysis, inline expansion of math functions, template optimizations.

File: llvm/lib/Target/ARM/ARMBaseInstrInfo.cpp
Comment (DragonDisciple) [inline llvm/lib/Target/ARM/ARMBaseInstrInfo.cpp:6829]: You could save some target code by making this unreachable a default implementation in PipelineLoopInfo, such that any target which enables MVE would by default run into the error case if they did not define it themselves.

This is similar to what's done in, for example, optimizeSelect (which is guarded by analyzeSelect), and replaceBranchWithTailCall (guarded by canMakeTailCallConditional)

File: llvm/include/llvm/CodeGen/TargetInstrInfo.h
Comment (DragonDisciple) [inline llvm/include/llvm/CodeGen/TargetInstrInfo.h:771]: "The all instructions" -> "All instructions"

File: llvm/include/llvm/CodeGen/TargetInstrInfo.h
Comment (DragonDisciple) [inline llvm/include/llvm/CodeGen/TargetInstrInfo.h:801]: Should be a sane default (false, almost certainly for now) to avoid forcing downstream targets to update their implementation.

File: llvm/lib/CodeGen/ModuloSchedule.cpp
Comment (DragonDisciple) [inline llvm/lib/CodeGen/ModuloSchedule.cpp:2139]: MachineBasicBlock::replacePhiUsesWith

File: llvm/lib/CodeGen/ModuloSchedule.cpp
Comment (DragonDisciple) [inline llvm/lib/CodeGen/ModuloSchedule.cpp:2134]: MachineBasicBlock::replaceSuccessor does the remove/add, plus more. Would it make sense to call that instead?

File: llvm/lib/CodeGen/ModuloSchedule.cpp
Comment (DragonDisciple) [inline llvm/lib/CodeGen/ModuloSchedule.cpp:2162]: Does this branch order make sense for optimal performance? If the 'expected' case is to run the pipelined loop, avoiding a forced branch seems more profitable, and the conditional should fire if and only if the 'exceptional' case of going around the pipelined loop occurs.

This is what the base expander seems to do, but it's confusing...

https://github.com/llvm/llvm-project/blob/300340f656d762afa8bde5fc398757d2951560bf/llvm/lib/CodeGen/ModuloSchedule.cpp#L878

It gets the trip count condition, and then inserts a branch that uses 'epilog' (branch-around) as the true and 'lastpro' (execute loop) as the false branch. This is counterintuitive but leads to a forced branch only if there's not enough iterations.

Unless I'm missing something, of course... I suppose a target could identify and swap the condition.

**Comment by DragonDisciple** [comment]
I am by no means an expert in AArch64, so someone with a better handle on it than I should review the implementations and tests.

However, this looks fine, and the results seem generally positive. I'd like to get this upstream so my team can pull it down and play with it. One thing I think we'd like to improve (and share upstream, of course!) is handling loop-carried dependences such as accumulations.

We have a pre-pipeliner unroller that does this, but think that using this method is likely cleaner.

tl;dr: LGTM

File: llvm/lib/Target/ARM/ARMBaseInstrInfo.cpp
Comment (ytmukai) [reply llvm/lib/Target/ARM/ARMBaseInstrInfo.cpp]: Fixed to provide default implementation. I previously considered it a convention not to provide a default implementation, but since the implementation of this interface is optional, it would be more convenient to have a default.

File: llvm/include/llvm/CodeGen/TargetInstrInfo.h
Comment (ytmukai) [reply llvm/include/llvm/CodeGen/TargetInstrInfo.h]: Fixed

File: llvm/include/llvm/CodeGen/TargetInstrInfo.h
Comment (ytmukai) [reply llvm/include/llvm/CodeGen/TargetInstrInfo.h]: Fixed to provide a default that returns false.

File: llvm/lib/CodeGen/ModuloSchedule.cpp
Comment (ytmukai) [reply llvm/lib/CodeGen/ModuloSchedule.cpp]: Fixed

File: llvm/lib/CodeGen/ModuloSchedule.cpp
Comment (ytmukai) [reply llvm/lib/CodeGen/ModuloSchedule.cpp]: Fixed. Thanks, I didn't know the appropriate interface.

File: llvm/lib/CodeGen/ModuloSchedule.cpp
Comment (ytmukai) [reply llvm/lib/CodeGen/ModuloSchedule.cpp:2184]: For now, I added a private option to swap the target blocks.

https://github.com/llvm/llvm-project/blob/594e3deb4c4501cfac6ebde266474f92226976c0/llvm/lib/CodeGen/ModuloSchedule.cpp#L25-L27

https://github.com/llvm/llvm-project/blob/594e3deb4c4501cfac6ebde266474f92226976c0/llvm/lib/CodeGen/ModuloSchedule.cpp#L2159-L2167

I was not aware of this, but if a processor predicts conditions, I think the number of executed instructions will be smaller if the 'expected' case is the destination of the conditional branch. But for processors that don't, I agree with you.
I would like to later create a framework for setting this up for each target, along with several other private options such as MVE activation.

**Comment by ytmukai** [comment]
Thanks for the review! I would be very grateful if you would use and improve this.

I understand pre-pipeliner unroll and issues about accumulations as I commented in an earlier discussion. (https://discourse.llvm.org/t/implementing-modulo-variable-expansion-for-machinepipeliner/71748/7?u=ytmukai)

As other improvements, we are aware of a problem of large unroll counts, which occur when the same addressing register is used by a load and a store. They are often located at the beginning and end of the DDG and cross all stages.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:9910]: Do we need to check for other uses of NZCV?

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:9830]: Does it need to be careful about other uses/defs of the NZCV flags?

**Review by davemgreen — COMMENTED** [review]
Sorry for the delay. I looked through the AArch64 code and it looks OK to me.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp
Comment (ytmukai) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.cpp]: Fixed. There was no reason to make it a pointer.

File: llvm/test/CodeGen/AArch64/sms-acceptable-loop3.mir
Comment (ytmukai) [reply llvm/test/CodeGen/AArch64/sms-acceptable-loop3.mir:1]: I omitted CBZ because of the additional processing required to support it and because we did not see it used in typical loops to pipeline. If that support is found to be useful I would like to modify it.

The tests were changed to remain and marked as unsupported.
https://github.com/llvm/llvm-project/blob/004574d8fbe851ed10f651ff7e99919f40b0ff95/llvm/test/CodeGen/AArch64/sms-acceptable-loop4.mir#L4-L5

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp
Comment (ytmukai) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:9931]: CALL instructions could cause problems because NZCV use/def is not set. It is fixed not to pipeline when CALL or instructions which have unmodeled side effects exist as https://github.com/llvm/llvm-project/pull/65609/commits/0afa899364a4eec2a514984801cb4a740655ee25.

Looking at the code in MachineScheduler, it seems that there are no other instructions that implicitly use NZCV.

https://github.com/llvm/llvm-project/blob/083a26682d1c4d2d6efb051cb2f658564eaa32ff/llvm/lib/CodeGen/ScheduleDAGInstrs.cpp#L550-L555

**Comment by ytmukai** [comment]
Thank you for the review! I think I have fixed all the issues.

File: llvm/test/CodeGen/AArch64/sms-acceptable-loop3.mir
Comment (davemgreen) [reply llvm/test/CodeGen/AArch64/sms-acceptable-loop3.mir:1]: Sounds good.

I would say it might be best to update the CHECK below instead or marking it unsupported. We can then check we don't do anything bad with the CBZ. The comment is good to keep in explaining that the instruction isn't supported.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp
Comment (davemgreen) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:9931]: Thanks.

**Review by davemgreen — APPROVED** [review]
Thanks. I would recommend updating the tests as opposed to marking them as unsupported, but otherwise LGTM.

File: llvm/test/CodeGen/AArch64/sms-acceptable-loop3.mir
Comment (ytmukai) [reply llvm/test/CodeGen/AArch64/sms-acceptable-loop3.mir:1]: Thank you. Fixed to make sure it is not pipelined.

**Comment by ytmukai** [comment]
Thank you for the approval!
The tests were fixed: https://github.com/llvm/llvm-project/pull/65609/commits/17609277d3402ad7ad0a19bad18ec4ded4b964b5

**Pull request merged**
